### ISSUE TITLE
segmentation fault on cv::morphologyEx on large images

### ISSUE THREAD

**Issue by hovnatan** [opencv/opencv#11296]
##### System information (version)

- OpenCV => 3.4.1
- Operating System / Platform => Ubuntu 18.04 x64
- Compiler => gcc 7.3.0

##### Detailed description

Segmentation fault on `cv::morphologyEx` on large, e.g., `25000x25000`, image.

##### Steps to reproduce
This program 

    int main() {
      cv::Mat1f m(cv::Size(25000, 25000), 0);
      cv::Mat kernel_closing = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5));
      cv::Mat result_mask;
      cv::morphologyEx(m, result_mask, cv::MORPH_CLOSE, kernel_closing);
      return 0;
    }

segmentation faults with this stack:
![screenshot from 2018-04-11 15-57-34](https://user-images.githubusercontent.com/8426582/38615498-efa3aa62-3da1-11e8-9f3c-29f07c9c3526.png)

**Comment by alalek** [comment]
Confirmed.
Workaround is to turn off IPP AVX optimizations via environment variable `OPENCV_IPP=sse42` or disable IPP completely: `OPENCV_IPP=disabled`

/cc @lupustr3

**Comment by alalek** [comment]
Morphology operations itself must be bit-exact (result should not be affected by optimization libraries or running on different platforms like x86/ARM).

But in general, the most part of computer vision operations are not bit-exact, if they uses somehow floating-points internally (rounding rules, errors).
Examples of such operations are: resize, floating-point filters, cvtColor (especially Lab, XYZ and other color spaces) and many others.

Also you could try to disable IPP for this call only (thread-local change):
```
bool useIPP = cv::ipp::useIPP();
cv::ipp::setUseIPP(false);
... morphology call ...
if (useIPP) cv::ipp::setUseIPP(true);
```

**Comment by lupustr3** [comment]
@hovnatan @alalek Thanks for the report. We will investigate this issue in IPP.

**Comment by madhusudangr** [comment]
System information (version)
OpenCV => 3.4.1
Operating System / Platform => Windows 10 64bit
Compiler => MSVC

Is mine the same error ? I get a similar segmentation fault at morphology operation, my call stack is as follows,

	bulls-eye.exe!cv::error(class cv::Exception const &)	C++
 	bulls-eye.exe!cv::error(int,class cv::String const &,char const *,char const *,int)	C++
 	bulls-eye.exe!cv::MorphFilter<struct cv::MinOp<unsigned char>,struct cv::MorphIVec<struct cv::VMin8u> >::MorphFilter<struct cv::MinOp<unsigned char>,struct cv::MorphIVec<struct cv::VMin8u> >(class cv::Mat const &,class cv::Point_<int>)	C++
 	bulls-eye.exe!cv::getMorphologyFilter(int,int,class cv::_InputArray const &,class cv::Point_<int>)	C++
 	bulls-eye.exe!cv::createMorphologyFilter(int,int,class cv::_InputArray const &,class cv::Point_<int>,int,int,class cv::Scalar_<double> const &)	C++
 	bulls-eye.exe!cv::morphologyEx(class cv::_InputArray const &,class cv::_OutputArray const &,int,class cv::_InputArray const &,class cv::Point_<int>,int,int,class cv::Scalar_<double> const &)	C++
 	bulls-eye.exe!cv::hal::morph(int,int,int,unsigned char *,unsigned __int64,unsigned char *,unsigned __int64,int,int,int,int,int,int,int,int,int,int,int,unsigned char *,unsigned __int64,int,int,int,int,int,double const * const,int,bool)	C++
 	bulls-eye.exe!cv::erode(class cv::_InputArray const &,class cv::_OutputArray const &,class cv::_InputArray const &,class cv::Point_<int>,int,int,class cv::Scalar_<double> const &)	C++
	bulls-eye.exe!using_HSV_color_space(cv::Mat image, bool display, bool save) Line 629	C++
 	bulls-eye.exe!fit_circles_check(cv::Mat image) Line 1305	C++
 	bulls-eye.exe!option8(std::basic_string<char,std::char_traits<char>,std::allocator<char> > _path) Line 2871	C++
 	bulls-eye.exe!test_fit_circles() Line 3005	C++
 	bulls-eye.exe!main() Line 3023	C++
 	[External Code]

The error on the console is `OpenCV(3.4.1) Error: Assertion failed (_kernel.type() == 0) in cv::MorphFilter<struct cv::MinOp<unsigned char>,struct cv::MorphIVec<struct cv::VMin8u> >::MorphFilter, file c:\opencv\opencv-3.4.1\modules\imgproc\src\morph.cpp, line 787`

I am not able to solve the error by using the solution provided in this thread

**Comment by alalek** [comment]
@madhusudangr No, your problem is not related to this issue. Error massage says that you should pass kernel with CV_8UC1 type.

Usage questions should go to Users OpenCV Q/A forum: http://answers.opencv.org

P.S. I will remove these two last messages from this issue tomorrow.

**Comment by maleficxp** [comment]
I also was faced with the same problem. 
Python code sample:
```
import cv2
import numpy as np

mask1 = np.zeros((46341, 46341), np.uint8)
mask2 = np.zeros((46342, 46342), np.uint8)
kernel = np.ones((3, 3), np.uint8)

mask1 = cv2.erode(mask1, kernel, iterations=1)
# works well
print("erode 1 complete") 

mask2 = cv2.erode(mask2, kernel, iterations=1)
# Segmentation fault
print("erode 2 complete")
```
Compiling opencv without IPP solved problem for me.
